Lib/idlelib/pyshell.py

```diff
@@ -1221,32 +1285,35 @@ def recall(self, s, event):
             if prefix.rstrip().endswith(':'):
                 self.newline_and_indent_event(event)
                 prefix = self.text.get("insert linestart", "insert")
-            self.text.insert("insert", lines[0].strip())
+            self.text.insert("insert", lines[0].strip(),
+                             self.user_input_insert_tags)
             if len(lines) > 1:
                 orig_base_indent = re.search(r'^([ \t]*)', lines[0]).group(0)
                 new_base_indent  = re.search(r'^([ \t]*)', prefix).group(0)
                 for line in lines[1:]:
                     if line.startswith(orig_base_indent):
                         # replace orig base indentation with new indentation
                         line = new_base_indent + line[len(orig_base_indent):]
-                    self.text.insert('insert', '\n'+line.rstrip())
+                    self.text.insert('insert', '\n' + line.rstrip(),
+                                     self.user_input_insert_tags)
         finally:
             self.text.see("insert")
             self.text.undo_block_stop()
 
     def runit(self):
+        index_before = self.text.index("end-2c")
         line = self.text.get("iomark", "end-1c")
         # Strip off last newline and surrounding whitespace.
         # (To allow you to hit return twice to end a statement.)
-        i = len(line)
-        while i > 0 and line[i-1] in " \t":
-            i = i-1
-        if i > 0 and line[i-1] == "\n":
-            i = i-1
-        while i > 0 and line[i-1] in " \t":
-            i = i-1
-        line = line[:i]
-        self.interp.runsource(line)
+        line = re.sub(r"[ \t]*\n?[ \t]*$", "", line)
+        input_is_complete = self.interp.runsource(line)
+        if input_is_complete or input_is_complete is None:
+            if (
+                self.user_input_insert_tags and
```

Comment:
In Shell, self.user_input_insert_tabs is the 'True' constant 'shell', making this test irrelevant.

I have no idea why '\n' needs to be untagged.

Reply:
True.

This PR makes the application of "input" tag ranges more consistent, due to inputs no longer being split by prompts in the text widget. Other features, such as recall, rely on "input" tag ranges to work properly.